synthesizer/program/src/resources/credits.aleo

```diff
@@ -106,573 +149,515 @@ record credits:
+        // Enforce that the number of delegators is less than or equal to 100_000.
+        assert.eq r10 true;
+        // Set the new number of delegators.
+        set r9 into metadata[aleo1qgqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqanmpl0];
+    // }
+
     position continue_bond_delegator;
 
     /* Bonded */
 
     // Construct the initial bond state.
-    cast r1 0u64 into r28 as bond_state;
+    cast r1 0u64 into r11 as bond_state;
     // Get the bond state for the caller, or default to the initial bond state.
-    get.or_use bonded[r0] r28 into r29;
+    get.or_use bonded[r0] r11 into r12;
     // Enforce the validator matches in the bond state.
-    assert.eq r29.validator r1;
+    assert.eq r12.validator r1;
 
     // Increment the microcredits in the bond state.
-    add r29.microcredits r3 into r30;
+    add r12.microcredits r3 into r13;
+
     // Determine if the amount is at least 10 thousand credits.
-    gte r30 10_000_000_000u64 into r31;
+    gte r13 10_000_000_000u64 into r14;
     // Enforce the amount is at least 10 thousand credits.
-    assert.eq r31 true;
+    assert.eq r14 true;
 
     // Construct the updated bond state.
-    cast r1 r30 into r32 as bond_state;
+    cast r1 r13 into r15 as bond_state;
 
     /* Account */
 
     // Get the balance of the caller.
     // If the account does not exist, this finalize scope will fail.
-    get account[r0] into r33;
+    get account[r0] into r16;
     // Decrement the balance of the caller.
-    sub r33 r3 into r34;
+    sub r16 r3 into r17;
 
     /* Writes */
 
-    // Update the committee state for the specified validator.
-    set r23 into committee[r1];
     // Update the bond state for the caller.
-    set r32 into bonded[r0];
+    set r15 into bonded[r0];
     // Update the balance of the caller.
-    set r34 into account[r0];
-
-    // The terminus.
-    position end;
+    set r17 into account[r0];
+    // Update total delegated amount.
+    // Get current total delegated amount.
+    get.or_use delegated[r1] 0u64 into r18;
+    // Add new bond amount to current delegation.
+    add r3 r18 into r19;
+    // Update delegated amount.
+    set r19 into delegated[r1];
 
 /**********************************************************************************************************************/
 
-// This function allows any staker to unbond their microcredits from a validator.
-// The corresponding function for 'unbond_public' is 'claim_unbond_public'.
+// Unbond a staker using their associated withdrawal address
+// **Validators**: It will remove the validator if the self bonded balance falls below 100 credits
+// or the total delegated balance falls below 10M credits.
+// **Delegators**: It will remove the validator if the total delegated balance falls below 10M credits.
+// It will remove the entire bond_state of the delegator if it falls below 10,000 credits.
+//
+// Validators can also forcibly unbond a delegator
+
 function unbond_public:
```

Comment:
If we are separating `bond_public` into `bond_validator` and `bond_delegator`.

Why not do the same for `unbond_public`? `unbond_validator` and `unbond_delegator` may be easier to implement and read, since we are already going down the route of separation.